docs/tutorial/Tutorial.md

```diff
@@ -0,0 +1,501 @@
+    use crate::vm::VMBinding;
+    use crate::MMTK;
+    use std::marker::PhantomData;
+    use std::ops::{Deref, DerefMut};
+    ```
+3. Add a new structure, `MyGCCopyContext`, with the type parameter `VM: VMBinding`. It should have the fields `plan:&'static MyGC<VM>` and `mygc: BumpAllocator`.
+4. Create an implementation block - `impl<VM: VMBinding> CopyContext for MyGCCopyContext<VM>`.
+   1. Add a type alias for VMBinding (given to the class as `VM`): `type VM: VM`. 
+   2. Add the following skeleton functions (taken from `plan/global.rs`):
+       ```rust
+       fn new(mmtk: &'static MMTK<Self::VM>) -> Self { };
+       fn init(&mut self, tls: OpaquePointer) { };
+       fn prepare(&mut self) { };
+       fn release(&mut self) { };
+       fn alloc_copy(`init
+           &mut self,
+           original: ObjectReference,
+           bytes: usize,
+           align: usize,
+           offset: isize,
+           semantics: AllocationSemantics,
+       ) -> Address {
+       };
+       fn post_copy(
+           &mut self,
+           _obj: ObjectReference,
+           _tib: Address,
+           _bytes: usize,
+           _semantics: AllocationSemantics,
+       ) {
+       }
+       ```
+   3. To `new`, add an initialiser for the class:
+       ```rust
+       Self {
+             plan: &mmtk.plan,
+             mygc: BumpAllocator::new(OpaquePointer::UNINITIALIZED, None, &mmtk.plan),
+         }
+       ```
+   4. In `init`, set the `tls` variable in the held instance of `mygc` to the one passed to the function.
+   5. In `prepare`, rebind the allocator to the tospace.
+   6. Leave `release` with an empty body.
+   7. In `alloc`, call the allocator's `alloc` function. Above the function, use an inline attribute (`#[inline(always)]`) to tell the Rust compiler to always inline the function.
+   8. In `post_copy` add the following code. Also, add an inline (always) attribute.
+       ```rust
+       forwarding_word::clear_forwarding_bits::<VM>(obj);
+       ```
+5. Add a new public structure, `MyGCProcessEdges`, with the type parameter `<VM:VMBinding>`. It will hold an instance of `ProcessEdgesBase` and `PhantomData`, and implement the Default trait:
+    ```rust
+    #[derive(Default)]
+    pub struct MyGCProcessEdges<VM: VMBinding> {
+        base: ProcessEdgesBase<MyGCProcessEdges<VM>>,
+        phantom: PhantomData<VM>,
+    }
+    ```
+6. Add a new implementations block `impl<VM:VMBinding> ProcessEdgesWork for MyGCProcessEdges<VM>`.
+   1. Add a VM type alias (`type VM = VM`).
+   2. Add a new method, `new`.
+       ```rust
+       fn new(edges: Vec<Address>, _roots: bool) -> Self {
+           Self {
+               base: ProcessEdgesBase::new(edges),
+               ..Default::default()
+           }
+       }
+      ```
+   3. Add a new method, `trace_object(&mut self, object: ObjectReference)`.
+     1. This method should return an ObjectReference, and use the inline (*not* always) attribute.
```

Comment:
We dont need to emphasis 'not always'. We are not sure the performance difference between `inline` and `inline(always)` here.